docs/development/design/build-images.rst

```diff
@@ -0,0 +1,274 @@
+
+   using ``Dockerfile.custom`` that has the following content:
+
+   .. code:: Dockerfile
+
+      # Dockerfile.custom
+      ARG RTD_IMAGE
+      FROM readthedocs/build:${RTD_IMAGE}
+
+      ARG RTD_NODE_VERSION
+      ARG RTD_RUST_VERSION
+      ARG RTD_APT_PACKAGES
+
+      USER root
+      WORKDIR /
+
+      # Install extras
+      RUN apt-get update
+      RUN apt-get install -y ${RTD_APT_PACKAGES}
+
+      USER docs
+      WORKDIR /home/docs
+
+      # Install ``nodejs``
+      RUN nodenv install ${RTD_NODE_VERSION}
+      RUN nodenv global ${RTD_NODE_VERSION}
+
+      # Install ``rust``
+      RUN curl https://sh.rustup.rs -sSf | sh -s -- -y --default-toolchain ${RTD_RUST_VERSION}
+      ENV PATH="/home/docs/.cargo/bin:$PATH"
+
+   Building this image should be pretty fast (~2 minutes locally) since all the requirements to install these extra packages
+   are already installed and all of them are pre-compiles binaries. It will take the time it takes to download them.
+
+
+Updating versions over time
+---------------------------
+
+How do we add/upgrade a Python version?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Python patch versions can be upgraded on the affected image.
+As the ``base`` image won't change for this case, it will only modify the layers after it.
+All the OS package versions will remain the same.
+
+In case we need to *add* a new Python version, we just need to build a new image based on ``base``:
+``ubuntu20-py310`` that will contain Python 3.10 and none of the other images are affected.
+
+
+How do we upgrade system versions?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+We usually don't upgrade these dependencies unless we upgrade the Ubuntu version.
+So, they will be only upgraded when we go from Ubuntu 18.04 LTS to Ubuntu 20.04 LTS for example.
+
+Examples of these versions are:
+
+* doxygen
+* git
+* subversion
+* pandoc
+* swig
+* latex
+
+This case will introduce a new ``base`` image. Example, ``ubuntu22-base`` in 2022.
+Note that these images will be completely isolated from the rest and don't require them to rebuild.
+This also allow us to test new Ubuntu versions without breaking people's builds.
+
+How do we add an extra requirement?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In case we need to add an extra requirement to the ``base`` image,
+we will need to rebuild all of them.
+The new image may have different package versions since there may be updates on the Ubuntu repositories.
+This conveys some small risk here, but in general we shouldn't require to add packages to the base images.
+
+Users with specific requirements could use ``build.apt`` and/or ``build.extras`` in the config file.
+
+How do we remove an old Python version?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+At some point an old version of Python will be deprecated (eg. 3.4) and will be removed.
+To achieve this, we can just remove the Docker image affected: ``ubuntu20-py34``,
+once there are no users depending on it anymore.
```

Comment:
I would consider raising some sort of warning in the build that surfaces on the build page. We shouldn't support these old unsupported versions forever just because somebody has forgotten they're using it.